From f58c7e4fd05ec13383518ccd51663167d45e92d0 Mon Sep 17 00:00:00 2001
From: Daniel Bylinka <daniel.bylinka@gmail.com>
Date: Fri, 2 Apr 2021 19:02:58 +0200
Subject: [PATCH] [statuscmd] Signal mouse button and click location to status
 monitor (CUSTOM MODIFIED for statuscolors compatibility)

Modified to use byte range 0x10-0x1F instead of 0x01-0x1F to avoid
conflicts with statuscolors patch (which uses 0x01-0x0C).

---
 config.def.h |   6 +++-
 dwm.c        | 100 ++++++++++++++++++++++++++++++++++++++++++++++++---
 2 files changed, 100 insertions(+), 6 deletions(-)

diff --git a/config.def.h b/config.def.h
index 9efa774..0b8b310 100644
--- a/config.def.h
+++ b/config.def.h
@@ -104,7 +104,9 @@ static Button buttons[] = {
 	{ ClkLtSymbol,		0,		Button1,	setlayout,	{0} },
 	{ ClkLtSymbol,		0,		Button3,	setlayout,	{.v = &layouts[2]} },
 	{ ClkWinTitle,		0,		Button2,	zoom,		{0} },
-	{ ClkStatusText,	0,		Button2,	spawn,		{.v = termcmd } },
+	{ ClkStatusText,	0,		Button1,	sigstatusbar,	{.i = 1} },
+	{ ClkStatusText,	0,		Button2,	sigstatusbar,	{.i = 2} },
+	{ ClkStatusText,	0,		Button3,	sigstatusbar,	{.i = 3} },
 	{ ClkClientWin,		MODKEY,		Button1,	movemouse,	{0} },
 	{ ClkClientWin,		MODKEY,		Button2,	togglefloating, {0} },
 	{ ClkClientWin,		MODKEY,		Button3,	resizemouse,	{0} },
diff --git a/dwm.c b/dwm.c
index 30070e5..a8e7d22 100644
--- a/dwm.c
+++ b/dwm.c
@@ -211,6 +211,7 @@ static void tagmon(const Arg *arg);
 static void tile(Monitor *m);
 static void togglebar(const Arg *arg);
 static void togglefloating(const Arg *arg);
+static void sigstatusbar(const Arg *arg);
 static void toggletag(const Arg *arg);
 static void toggleview(const Arg *arg);
 static void unfocus(Client *c, int setfocus);
@@ -328,6 +329,9 @@ struct Pertag {
 };

 static unsigned int scratchtag = 1 << LENGTH(tags);
+static int statusw;
+static int statussig;
+static pid_t statuspid = -1;

 /* compile-time check if all tags fit into an unsigned int bit array. */
 struct NumTags { char limitexceeded[LENGTH(tags) > 31 ? -1 : 1]; };
@@ -367,12 +371,30 @@ buttonpress(XEvent *e)
 		else if (ev->x < x + blw)
 			click = ClkLtSymbol;
-		else if (ev->x > selmon->ww - (int)TEXTW(stext))
+		else if (ev->x > selmon->ww - statusw) {
+			x = selmon->ww - statusw;
 			click = ClkStatusText;
-		else
+			statussig = 0;
+			for (text = s = stext; *s && x <= ev->x; s++) {
+				if ((unsigned char)(*s) >= 0x10 && (unsigned char)(*s) <= 0x1f) {
+					ch = *s;
+					*s = '\0';
+					x += TEXTW(text) - lrpad;
+					*s = ch;
+					text = s + 1;
+					if (x >= ev->x)
+						break;
+					statussig = ch;
+				}
+			}
+		} else
 			click = ClkWinTitle;
 	} else if ((c = wintoclient(ev->window))) {
 		focus(c);

+	char *text, *s, ch;
+
 	switch(click) {

@@ -605,9 +627,24 @@ drawbar(Monitor *m)

 	/* draw status first so it can be overdrawn by tags later */
 	if (m == selmon) { /* status is only drawn on selected monitor */
+		char *text, *s, ch;
 		drw_setscheme(drw, scheme[SchemeNorm]);
-		tw = TEXTW(stext) - lrpad + 2; /* 2px right padding */
-		drw_text(drw, m->ww - tw, 0, tw, bh, 0, stext, 0);
+
+		x = 0;
+		for (text = s = stext; *s; s++) {
+			if ((unsigned char)(*s) >= 0x10 && (unsigned char)(*s) <= 0x1f) {
+				ch = *s;
+				*s = '\0';
+				tw = TEXTW(text) - lrpad;
+				drw_text(drw, m->ww - statusw + x, 0, tw, bh, 0, text, 0);
+				x += tw;
+				*s = ch;
+				text = s + 1;
+			}
+		}
+		tw = TEXTW(text) - lrpad + 2;
+		drw_text(drw, m->ww - statusw + x, 0, tw, bh, 0, text, 0);
+		tw = statusw;
 	}

 	for (c = m->clients; c; c = c->next) {
@@ -1950,6 +1987,20 @@ spawn(const Arg *arg)
 	}
 }

+void
+sigstatusbar(const Arg *arg)
+{
+	union sigval sv;
+
+	if (!statussig)
+		return;
+	sv.sival_int = arg->i;
+	if ((statuspid = getstatusbarpid()) <= 0)
+		return;
+
+	sigqueue(statuspid, SIGRTMIN+statussig, sv);
+}
+
 void
 toggletag(const Arg *arg)
 {
@@ -2088,10 +2139,27 @@ updatesizehints(Client *c)
 void
 updatestatus(void)
 {
-	if (!gettextprop(root, XA_WM_NAME, stext, sizeof(stext)))
+	if (!gettextprop(root, XA_WM_NAME, stext, sizeof(stext))) {
 		strcpy(stext, "dwm-"VERSION);
+		statusw = TEXTW(stext) - lrpad + 2;
+	} else {
+		char *text, *s, ch;
+
+		statusw  = 0;
+		for (text = s = stext; *s; s++) {
+			if ((unsigned char)(*s) >= 0x10 && (unsigned char)(*s) <= 0x1f) {
+				ch = *s;
+				*s = '\0';
+				statusw += TEXTW(text) - lrpad;
+				*s = ch;
+				text = s + 1;
+			}
+		}
+		statusw += TEXTW(text) - lrpad + 2;
+
+	}
 	drawbar(selmon);
 }

 void
 unfocus(Client *c, int setfocus)
--
2.31.0
